File: app/src/main/java/com/happs/ximand/ringcontrol/model/object/command/ReplaceTimetableCommand.java
package com.happs.ximand.ringcontrol.model.object.command;

import com.happs.ximand.ringcontrol.model.object.timetable.Lesson;
import com.happs.ximand.ringcontrol.model.object.timetable.Time;
import com.happs.ximand.ringcontrol.model.object.timetable.Timetable;

public class ReplaceTimetableCommand extends BluetoothCommand {

    private static final byte REPLACE_TIMETABLE_COMMAND_CODE = 10;
    private static final byte TIMETABLE_ID_STUB = 1;

    private final byte[] command;
    private final Timetable timetable;

    public ReplaceTimetableCommand(Timetable newTimetable) {
        super(REPLACE_TIMETABLE_COMMAND_CODE);
        int timeSize = newTimetable.getLessons().size() * 2 * 3;
        this.command = new byte[3 + timeSize + 1];
        this.timetable = newTimetable;
    }

    @Override
    public byte[] getCommand() {
        appendMeta();
        for (int i = 0; i < timetable.getLessons().size(); i++) {
            Lesson lesson = timetable.getLessons().get(i);
            appendTime(i * 6 + 3, lesson.getStartTime());
            appendTime(i * 6 + 6, lesson.getEndTime());
        }
        appendEndOfLine();
        return command;
    }

    private void appendMeta() {
        command[0] = getCode();
        command[1] = TIMETABLE_ID_STUB;
        command[2] = (byte) (timetable.getLessons().size() * 2);
    }

    private void appendTime(int startPos, Time time) {
        command[startPos] = (byte) time.getHours();
        command[startPos + 1] = (byte) time.getMinutes();
        command[startPos + 2] = (byte) time.getSeconds();
    }

    private void appendEndOfLine() {
        command[command.length - 1] = 1;
    }
}

File: app/src/main/java/com/happs/ximand/ringcontrol/model/object/command/ReplaceTimetableCommand.kt
package com.happs.ximand.ringcontrol.model.`object`.command

import com.happs.ximand.ringcontrol.model.`object`.timetable.Lesson
import com.happs.ximand.ringcontrol.model.`object`.timetable.Time
import com.happs.ximand.ringcontrol.model.`object`.timetable.Timetable

class ReplaceTimetableCommand(newTimetable: Timetable) : BluetoothCommand<Void>(REPLACE_TIMETABLE_COMMAND_CODE) {

    private val command: ByteArray
    private val timetable: Timetable

    override fun toByteArray(): ByteArray {
        appendMeta()
        for (i in timetable.lessons.indices) {
            val lesson: Lesson = timetable.lessons[i]
            appendTime(i * 6 + 3, lesson.startTime)
            appendTime(i * 6 + 6, lesson.endTime)
        }
        appendEndOfLine()
        return command
    }

    override fun getMainContent(): Void {
        throw UnsupportedOperationException()
    }

    private fun appendMeta() {
        command[0] = code
        command[1] = TIMETABLE_ID_STUB
        command[2] = (timetable.lessons.size * 2).toByte()
    }

    private fun appendTime(startPos: Int, time: Time) {
        command[startPos] = time.hours.toByte()
        command[startPos + 1] = time.minutes.toByte()
        command[startPos + 2] = time.seconds.toByte()
    }

    private fun appendEndOfLine() {
        command[command.size - 1] = 1
    }

    companion object {
        private const val REPLACE_TIMETABLE_COMMAND_CODE: Byte = 11
        private const val TIMETABLE_ID_STUB: Byte = 1
    }

    init {
        val timeSize: Int = newTimetable.lessons.size * 2 * 3
        command = ByteArray(3 + timeSize + 1)
        timetable = newTimetable
    }
}